adding header handling in routeVeam + unit tests

benzekrimaha opened the pull request:
Issue: CLDSRV-806

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

DarkIsDude commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
         // Deletion requires that the tags of the object are returned.
-        if (requestQueryParams && Object.keys(requestQueryParams).length > 0
-            && !(method === 'GET' && (requestQueryParams['X-Amz-Credential'] || ('tagging' in requestQueryParams)))) {
+        const queryKeys = Object.keys(requestQueryParams || {});
```
Here we parse 3 times the requestQueryParams with same logic (lowerCase then compare). We can maybe do eveything in one shot if this route is performance sensitive ?

SylvainSenechal replied:
Possibly this, and also considering an early return based on the length of queryKeys or requestQueryParams ?

DarkIsDude commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+
+        if (!isPresignedGet && queryKeys.length > 0) {
+            const unsupportedKeys = queryKeys
+                .filter(key => {
```
We can do a some and return directly instead of continuing the iteration ?

benzekrimaha replied:
indeed now as soon as we have an unsupportedKey we return

DarkIsDude commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+                    if (allowedSdkQueryKeys.has(lowerKey)) {
+                        return false;
+                    }
+                    if (lowerKey.startsWith('x-amz-sdk-')) {
```
Why this one and not the array at the top ? Why are we more permissive here ?

benzekrimaha replied:
The startsWith('x-amz-sdk-') check is a defensive fallback: AWS SDKs occasionally adds new telemetry keys under that namespace, and I wanted to avoid blocking a future patch release that introduces, say, x-amz-sdk-language if we’d rather force every key to be enumerated explicitly, I can drop the prefix guard and expand the array whenever we encounter a new one. let me know what you think

francoisferrand commented on .github/pykmip/Dockerfile:
```diff
@@ -4,22 +4,22 @@ FROM python:3.10-alpine
         python3-dev \
         libffi-dev \
-        libressl-dev \
+        openssl-dev \
```
why switch from `libressl` to `openssl` ?
AFAIU we actually want to use libressl, which is the more secure & modern lib...

SylvainSenechal commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+        const xIdValue = _getQueryValueCaseInsensitive(requestQueryParams, 'x-id');
+        if (xIdValue && xIdValue !== apiToAction[method]) {
             return errorInstances.InvalidRequest
                 .customizeDescription('The Veeam SOSAPI folder does not support this action.');
```
Is veeam as strict as AWS regarding errors messages, or can we customize it some more ? : 
"the x-id value you provided in the query parameter: xxx, is unexpected, it should be be xxx for this api"

benzekrimaha replied:
As nothing was specified I'd rather keep it that way in this PR , we can work on better error logging later on as today we only have this kind of errors

SylvainSenechal commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
         // Download relies on GETs calls with auth in query parameters, that can be
         // checked if 'X-Amz-Credential' is included.
         // Deletion requires that the tags of the object are returned.
```
This older comment is mentioning that this logic is used to allow certains apis to go through. 

It mentions the Deletion apis requiring tagging. But the logic later checks that the value of method is GET. Will this work when the method is DELETE ? I see in your tests you added 3 tests that are all PutObject, I say let's test it with a DeleteObject too

SylvainSenechal replied:
More general comment and follow up : I find it hard to reason about both the previous logic and the new one, as these logic try to fit all apis at the same time, with long if checking for method type, different headers. For example, the tagging check I only useful for the DeleteObjects if I understood correctly. 
What would happen if if later we had another constraint to deal with that only applies to one of the 5 apis   

Considering that we have a nice enum list "apiToAction" of all possible apis that this router will handle, I would add a function with an exhaustive switch statement to treat all 5 apis separately.

I also understand that we may not have time to do this

SylvainSenechal replied:
I think the comment about the delete is missleading, i think this comment is still talking about the Get api, and saying that the get method should return tags in order to be able to perform the deletion later. But the way and the place where its written makes it look like we want to allow DELETE apis. anyways

benzekrimaha replied:
Before we touched it, the route rejected every non-GET request that carried any query parameters.That’s why our PUT started failing as soon as the SDK v3 appended ?x-id=…. The change simply teaches the validator that x-id (and the handful of SDK telemetry keys) are acceptable, while keeping the old rule for everything else.

SylvainSenechal commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+                && (_hasQueryKeyCaseInsensitive(requestQueryParams, 'X-Amz-Credential')
+                    || _hasQueryKeyCaseInsensitive(requestQueryParams, 'tagging')));
+
+        if (!isPresignedGet && queryKeys.length > 0) {
```
With the previous logic, having !isPresignedGet && queryKeys.length > 0 was enough to return an error. 
Can we return an error directly here, and do the check unsupportedKeys right after, because here the 2 are mixed but I think it's not necessary ?

SylvainSenechal reviewed (changes requested):
Some comments, I think no need for a bigger refactoring, but some simplifications could be possible ? + adding a test with method different than GET

SylvainSenechal commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +89,32 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+                    }
+                    return true;
+                });
+            if (unsupportedKeys.length > 0) {
```
Wait so here, if the api ever receive any query different than the ones listed in allowedSdkQueryKeys, we will block the apis ? Isn't it risky because it forces us to be exhaustive in the list of allowed keys, are we sure this list is now complete ?

benzekrimaha replied:
Before we touched it, the route rejected every non-GET request that carried any query parameters.That’s why our PUT started failing as soon as the SDK v3 appended ?x-id=…. The change simply teaches the validator that x-id (and the handful of SDK telemetry keys) are acceptable, while keeping the old rule for everything else. For the list I tried to make it as exhaustive as possible with everything I could find on official documentations

SylvainSenechal commented on tests/unit/internal/routeVeeam.js:
```diff
@@ -77,6 +77,44 @@ describe('RouteVeeam: checkBucketAndKey', () => {
+        assert.strictEqual(err, undefined);
+    });
+
+    it('should allow AWS SDK auxiliary x-amz-sdk-* query params on PUT for system.xml', () => {
```
nit: could add a test with a wrong param x-iamwrong ?

benzekrimaha replied:
we're already testing a wrong value here => https://github.com/scality/cloudserver/pull/6022/commits/51810dd8fdf5181efc87e5056ff93223acc6e0e4#diff-7d9febf34141ec8f0c48580a4e9e899f15f34227d9ecc39fcda6c5392e87f581R108, isn't it over testing to add other headers as we already know that we reject them directly ?

francoisferrand replied:
looking at all the branches in the code, more tests are needed : e.g. having "presigned" headers, on GET or otherwise, different case for headers, ...

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/CLDSRV-806-handling-additionnal-headers` has diverged from
`development/9.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/9.0` into `improvement/CLDSRV-806-handling-additionnal-headers`
* Rebase `improvement/CLDSRV-806-handling-additionnal-headers` onto `origin/development/9.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

francoisferrand commented on .github/pykmip/Dockerfile:
```diff
@@ -13,13 +13,12 @@ RUN apk add --no-cache \
     cd pykmip && \
-    python3 setup.py install && \
+    pip3 install . && \
+    cd / && \
```
why do we need to `cd /` ?
everything afterwards uses absolute path...

francoisferrand commented on .github/pykmip/Dockerfile:
```diff
@@ -13,13 +13,12 @@ RUN apk add --no-cache \
     git clone https://github.com/openkmip/pykmip.git && \
     cd pykmip && \
-    python3 setup.py install && \
+    pip3 install . && \
```
what is the reason for the build failure? Is there a new release of pykmip?
should the fix not include specifying the version of pykmip that we checkout, so we avoid unexpected behavior changes (or incompatibility when it requires a newer version of python...) ?

benzekrimaha replied:
The build was failing in the python3 setup.py install step when building pykmip from the GitHub repo on python:3.10-alpine (the setup.py invocation was exiting with status 1). There isn’t a new PyPI release of pykmip; we’re building from the GitHub repo because the latest fixes we need are only there (the last tagged release is still 0.10.0).

francoisferrand replied:
if we need a specific SHA1, we can (and should!) specify it anyway; and add a comment explaining why we don't use the release / ...

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -29,6 +29,13 @@ const apiToAction = {
+    'x-id',
+    'x-amz-user-agent',
+    'x-amz-sdk-invocation-id',
+    'x-amz-sdk-request',
```
not needed, these are covered by the generic `lowerKey.startsWith('x-amz-sdk-')` condition you added

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +74,33 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+            && (normalizedQueryKeys.has('x-amz-credential')
+                || normalizedQueryKeys.has('tagging'));
+
+        if (!isPresignedGet && normalizedQueryKeys.size > 0) {
```
I guess "extra headers" should not be supported either with presigned URL
(looking at the earlier code, it was -slightly incorrectly- trying to check that there is no header unless it is `X-Amz-Credential` or `tagging` on GET)

....but indeed, there are quite a few extra headers to check in the presigned case (X-Amz-Expires, X-Amz-Signature... c.f. extractQueryParams in arsenal)

→ I wonder if it really makes sense to be strict only for non-presigned Get requests: either we should be strict in any case, or never ?

benzekrimaha replied:
Shud this be looked upon in this PR or rather create a ticket to check this later as we're not altering the previous behaviour here ?

francoisferrand replied:
we are altering the behavior anyway : if we decide to be more lenient, no reason to try to fix the behavior and be "sometimes strict but less strict than before" (which also risks breaking if the SDK "randomly" adds yet more headers we do not support)

The question is really : what are we protecting against with these headers? Are there specific headers that MUST NOT be passed?
(naively, I don't think there are headers when doing a s3 get object request which are expected to modify the semantics : so just ignoring headers should be fine.... unless Veeam "spec" requires otherwise)

francoisferrand replied:
so I guess we should double check the veeam spec ;
and using the same validation as regular get object : no more, no less...

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -67,8 +74,33 @@ function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, lo
+
+        if (!isPresignedGet && normalizedQueryKeys.size > 0) {
+            const hasUnsupportedKeys = Array.from(normalizedQueryKeys.keys())
+                .some(lowerKey => {
```
this is very inefficient: instead of just going through the list one, you
- create a new map (converting & storing every header in lowercase)
- build yet another array of the keys
- iterate over every entry + explicitly check if presigned and/or x-id

```suggestion
let presigned = false;
for [key, value] in Object.entries(requestQueryParams || {}) {
    const normalizedKey = key.toLowerCase();
    if(normalizedKey == "x-id" && value !== apiToAction[method]) {
        return errorInstances.InvalidRequest
            .customizeDescription('The Veeam SOSAPI folder does not support this action.');
    } 
    if (normalizedKey == 'x-amz-credential' || normalizedKey == 'tagging') {
        presigned = true;
    }
    if (normalizedKey != 'x-amz-user-agent' && !normalizedKey.startsWith('x-amz-sdk-')) {
        return errorInstances.InvalidRequest
            .customizeDescription('The Veeam SOSAPI folder does not support this action.');
    }
}
```

francoisferrand commented:
branch is not up-to-date with upstream development/9.0 : needs to be rebase and resolve conflicts...

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-806 contains:


* `9.1.6`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.34`

* `9.1.7`

* `9.2.9`


Please check the `Fix Version/s` of CLDSRV-806, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-806 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.34`

* `9.1.7`

* `9.2.9`


Please check the `Fix Version/s` of CLDSRV-806, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-806 contains:


* `9.0.34`

* `9.1.7`

* `9.2.9`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.34`

* `9.1.7`

* `9.2.10`


Please check the `Fix Version/s` of CLDSRV-806, or the target
branch of this pull request.

benzekrimaha committed:
adding header handling in routeVeam + unit tests

This commit aims to add the handling of additional headers
for Veeam routes in the CloudServer. The change actually enables
requests with automatically added headers by the sdk to not be directly
rejectded.
Issue: CLDSRV-806

benzekrimaha committed:
checking out on latest commit on the pykmip repo

The image build was failing because we were trying to install pykmip
with python v2 (setup.py install uses python command which points to python v2
by default in alpine image). This change checks out the latest commit and
installs pykmip with pip3 as this is the one now supported by pykmip.
Issue: CLDSRV-806

benzekrimaha committed:
project version bump

Issue: CLDSRV-806

benzekrimaha committed:
removing ceph tests from ci

Issue: CLDSRV-806

benzekrimaha commented:
/create_integration_branches

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-806 contains:


* `9.0.34`

* `9.1.7`

* `9.2.9`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.34`

* `9.1.7`

* `9.2.10`


Please check the `Fix Version/s` of CLDSRV-806, or the target
branch of this pull request.




*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

benzekrimaha commented:
/create_pull_requests

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-806-handling-additionnal-headers` into
`development/9.0`
* [w/9.1/improvement/CLDSRV-806-handling-additionnal-headers](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/improvement/CLDSRV-806-handling-additionnal-headers) will be merged into `development/9.1` (pull request #6029)
* [w/9.2/improvement/CLDSRV-806-handling-additionnal-headers](https://github.com/scality/cloudserver/compare/development/9.2...w/9.2/improvement/CLDSRV-806-handling-additionnal-headers) will be merged into `development/9.2` (pull request #6030)



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests, create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_pull_requests, create_integration_branches**

benzekrimaha commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.0`

* :heavy_check_mark: `development/9.1`

* :heavy_check_mark: `development/9.2`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`



Please check the status of the associated issue CLDSRV-806.

Goodbye benzekrimaha.




*The following options are set:* **approve, create_pull_requests, create_integration_branches**
